docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+configuration. A single replica is used, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
+
+### Element: Seed API Server Extension Service
+The prometheus-metrics-adapter is registered as an API extension to the seed API server. It is deployed with minimal
+deviations from the default configuration, to reduce future merge effort. As a result, it currently takes charge not only
+of the custom metric API, but also of the external metric API, preventing the deployment of alternative handlers for
+these API extension points. 
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+4. New feature flag promoted and eventually removed.
```

Comment:
What are the criteria for promoting the feature gate? What steps need to be performed?

I could imagine, that proper monitoring needs to be introduced that allows developers/operators to judge the quality of the new solution.

Reply:
I'm all for hard criteria for feature gate promotion, which concrete ideas do you have in mind to judge the quality of an autoscaling solution? 

Right now, I can come up with an idea about _efficiency_: how much of the requested resources of each instance is actually used, to ensure the resource requests are appropriately set for each instance. Understanding if the number of instances is set correctly seems to be harder, I don't know if there's a good metric other than "fewer instances is better".
Looking at _effectiveness_, this seems even harder to measure. We can watch out for an increase in rejected requests or increase in request latencies on the apiservers – but this will most likely be a case-by-case thing and seems hard to be done programmatically for all apiservers? Maybe there are other, better ways to understand the quality of an autoscaling solution?

Reply:
I have no specific formal criteria in mind, at this time. That would make the plan the same as with most changes:
- Appoint key metrics/tests (in this case, e..g. API availability and latency, resource allocations, resource consumption, allocation-consumption relation, etc.)
- Measure on a representative sampling set
- Explore results with focus on application fitness/QoS and resource-level efficiency.